### DISCUSSION TITLE
EC2 iPXE TFTP Boot Help Please

### DISCUSSION THREAD

**Discussion by Tian105** [ipxe/ipxe#650] [Q&A]
Hi guys, Sorry this might be a stupid question, but I cant see to get ipxe to work on EC2. Im trying to use ipxe to boot an image off of the NAS.

I have 2 instances on EC2 in the same subnet with open firewall. 

1 instance is the NAS with TFTP, DHCP server configured (192.168.122.235). 

the other instance is the ipxe AMI.

I have the dhcp server configured to point 192.168.122.60 to the mac address of the ipxe instance, below is the dhcp configuration file

ddns-update-style none;
default-lease-time 86400;
max-lease-time 604800;
authoritative;

subnet 192.168.122.0 netmask 255.255.255.0 {
        host aocd_vm1 {
                option host-name "aocd_vm1";
                hardware ethernet 02:4d:e0:6c:f8:d6;
                fixed-address 192.168.122.60;
                filename "/var/lib/tftpboot/pxelinux.0";
                option dhcp-server-identifier 192.168.122.235;
                        }
}

the directory of the netboot files of a ubuntu distro is under /var/lib/tftpboot/

I have modified the user data of the ipxe instance to the following:
  #!ipxe
  set server_ip 192.168.122.235
  set tftp_path /srv/tftp/ubuntu-18.04 -desktop-amd64
  kernel tftp://${server_ip}${tftp_path}/casper/vmlinuz.efi || read void
  initrd tftp://${server_ip}${tftp_path}/casper/initrd.lz || read void
  imgargs vmlinuz.efi initrd=initrd.lz root=/dev/tftp boot=casper netboot=tftp tftproot=${server_ip}:${tftp_path} ip=dhcp splash quiet - 
  -  || read void
  boot || read void

The ipxe instance will attempt to read the user data file and get stuck in a loop of attempting to boot and fails.

Any pointers would be greatly appreciated.

**Comment by stappersg** [comment]
* https://www.markdownguide.org/extended-syntax/#fenced-code-blocks

**Comment by NiKiZe** [comment]
What do you mean by loop? Have you tried adding echos for debug?
I would not use tftp at all (instead use http), and you also should not need pxelinux.

But what does the tftp logs show?

**Reply by stappersg** [reply]
The `tftp logs` will show the loop from _"stuck in a loop of attempting to boot"_.

Yeah, I have the nasty feeling that @NiKiZe and @Tian105 are unaware of https://ipxe.org/howto/chainloading#breaking_the_infinite_loop

**Reply by NiKiZe** [reply]
If you where aware what "user data" is and what it does you would not make that comment. Especially since you are fully aware about the iPXE loop in normal circumstances, but not relevant here since pxelinux is in the DHCP config.

**Comment by epacua** [comment]
First, this line:

`filename` "/var/lib/tftpboot/pxelinux.0";

 does not seem right. If the root of your tftp folder is /var/lib/tftpboot/tftpboot, then you just need to put 'pxelinux.0'.
 
 Second, have you tried grabbing the file via tftp from a running system? Or perhaps do some investigation on the host server like checking if tftpd is actually running, checking its log, see if firewall is accepting incoming connections to it etc.
 
 Lastly, you seem to have conflicting configuration files. So you are trying to grab pxelinux.0 and yet the script is ipxe. With pxelinux, config files are found at pxelinux.cfg.
